ranking/src/ts_system/mod.rs

```diff
@@ -0,0 +1,314 @@
+
+use std::rc::{Rc, Weak};
+use std::f64::INFINITY;
+
+type TSPlayerRating = Gaussian;
+type TSMessage = nodes::Message;
+type TSPlayer = String;
+type TSTeam = Vec<TSPlayer>;
+type TSContestPlace = Vec<TSTeam>;
+type TSContest = Vec<TSContestPlace>;
+type TSRating = HashMap<TSPlayer, TSPlayerRating>;
+
+// TrueSkillStPB rating system
+pub struct TrueSkillSPBSystem {
+    // epsilon used for ties
+    eps: f64,
+    // default player rating
+    mu: f64,
+    // default player sigma
+    sigma: f64,
+    // performance sigma
+    beta: f64,
+    // epsilon used for convergence loop
+    convergence_eps: f64,
+    // defines sigma growth per second
+    sigma_growth: f64,
+    // history of ratings for all players
+    rating: TSRating,
+}
+
+impl Default for TrueSkillSPBSystem {
+    fn default() -> Self {
+        Self {
+            eps: 0.70,
+            mu: 1500.,
+            sigma: 1500. / 3., // mu/3
+            beta: 1500. / 6., // sigma/2
+            convergence_eps: 2e-4,
+            sigma_growth: 5.,
+            rating: TSRating::new(),
+        }
+    }
+}
+
+// fn update_rating(old: &TSRating, new: &mut TSRatingHistory, contest: &TSContest, when: usize) {
+//     for place in &contest[..] {
+//         for team in &place[..] {
+//             for player in &team[..] {
+//                 new.entry(player.clone()).or_insert(Vec::new()).push((old.get(player).unwrap().clone(), when));
+//             }
+//         }
+//     }
+// }
+
+
+fn gen_team_message<T, K: Clone>(places: &Vec<Vec<T>>, default: &K) -> Vec<Vec<K>> {
+    let mut ret: Vec<Vec<K>> = Vec::with_capacity(places.len());
+
+    for place in places {
+        ret.push(vec![default.clone(); place.len()]);
+    }
+
+    ret
+}
+
+
+fn gen_player_message<T, K: Clone>(places: &Vec<Vec<Vec<T>>>, default: &K) -> Vec<Vec<Vec<K>>> {
+    let mut ret = Vec::with_capacity(places.len());
+
+    for place in places {
+        ret.push(Vec::with_capacity(place.len()));
+
+        for team in place {
+            ret.last_mut().unwrap().push(vec![default.clone(); team.len()]);
+        }
+    }
+
+    ret
+}
+
+
+fn infer1(who: &mut Vec<impl TreeNode>) {
+    for item in who {
+        item.infer();
+    }
+}
+
+
+fn infer2(who: &mut Vec<Vec<impl TreeNode>>) {
+    for item in who {
+        infer1(item);
+    }
+}
+
+
+fn infer3(who: &mut Vec<Vec<Vec<impl TreeNode>>>) {
+    for item in who {
+        infer2(item);
+    }
+}
+
+
+fn infer_ld(ld: &mut Vec<impl TreeNode>, l: &mut Vec<impl TreeNode>) {
+    for i in 0..ld.len() {
+        l[i].infer();
+        ld[i].infer();
+    }
+    l.last_mut().unwrap().infer();
+    for j in 0..ld.len() {
+        let i = ld.len() - 1 - j;
+        ld[i].infer();
+        l[i].infer();
+    }
+}
+
+
+fn check_convergence(a: &Vec<Rc<RefCell<(TSMessage, TSMessage)>>>,
+                     b: &Vec<(TSMessage, TSMessage)>) -> f64 {
+    if a.len() != b.len() {
+        return INFINITY;
+    }
+
+    let mut ret = 0.;
+
+    for i in 0..a.len() {

```

Comment:
Alternative ways to write this:
1.
```
ret = ret.max(asdf)
            .max(jkl)
            .max(qwerty);
````
2.
```
let ret = a.iter().flat_map(|a_i| {
    let a_i = RefCell::borrow(a_i);
    vec![asdf, jkl, qwerty])
}.something()
```
The `something()` is a bit tricky. Normally you'd do something like `.max().unwrap_or(f64::NEG_INFINITY)` (or just a plain `unwrap()` if `a` is guaranteed to be non-empty); however, floats might be NaN which aren't totally ordered, so this isn't allowed. Alternatives are:
2a. `.max_by(|a, b| a.partial_cmp(b).unwrap()).unwrap_or(f64::NEG_INFINITY)`
If you use the above comparator a lot, you can give it a name instead of making it anonymous. Or...
2b. `.fold(f64::NEG_INFINITY, f64::max)`

The difference between 2a and 2b is in how they handle NaNs: 2a would just crash (which is probably better in this case), whereas 2b would treat NaN as the lowest possible value, effectively ignoring them as long as a normal value exists.

P.S. Consider using `expect()` instead of `unwrap()` to make a custom error message.